smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpProtocolGeneratorUtils.java

```diff
@@ -238,4 +242,38 @@ static void generateCollectBodyString(GenerationContext context) {
 
         return errorShapes;
     }
+
+    /**
+     * Writes resolved hostname, prepending existing hostname with hostPrefix and replacing each hostLabel with
+     * the corresponding top-level input member value.
+     *
+     * @param context The generation context.
+     * @param operation The operation to generate for.
+     */
+    static void writeHostPrefix(
+            GenerationContext context,
+            OperationShape operation
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        EndpointTrait trait = operation.getTrait(EndpointTrait.class).get();
+        writer.write("let resolvedHostname = (context.endpoint as any).hostname;");
+        // Check if disableHostPrefixInjection has been set to true at runtime
+        writer.openBlock("if (context.disableHostPrefix !== true) {", "}", () -> {
+            writer.write("resolvedHostname = $S + resolvedHostname;", trait.getHostPrefix().toString());
+            List<Pattern.Segment> prefixLabels = trait.getHostPrefix().getLabels();
+            Shape inputShape = context.getModel().expectShape(operation.getInput().get());
+            for (Pattern.Segment label : prefixLabels) {
+                MemberShape target = inputShape.members().stream().filter(s -> s.getMemberName().equals(
+                        label.getContent())).findFirst().get();
+                String memberName = symbolProvider.toMemberName(target);
+                writer.write("resolvedHostname = resolvedHostname.replace($S, input.$L)",
+                        "{" + label.getContent() + "}", memberName);
+            }
+            writer.write("const hostPattern = /^[a-zA-Z0-9]{1}$$|^[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9]$$/;");
```

Comment:
Is there any way to make this a library method somewhere so that we don't need to recompile the regex each time an operation is invoked? It would also help remove this hairy regex from codegen code and move it into JS land.

Someline like a method like `isValidHostname(value)`

Reply:
added a validHostname function in util-valid-hostname: https://github.com/aws/aws-sdk-js-v3/pull/774